### PR TITLE
5307 search for medicationrequests with medication contained does not return correct results

### LINKED ISSUE #5307: Search for MedicationRequests with medication contained does not return correct results
**Describe the bug**
When a search is made to find a MedicationRequest with a certain Medication, the results will not return the correct MedicationRequest if the Medication is included as a contained resource. This is with IndexOnContainedResources set to true and using `_contained=true`.

**To Reproduce**
1. Enabled IndexOnContainedResources
2. Submit the following MedicationRequest resource
```
{
	"resourceType": "MedicationRequest",
	"contained": [
		{
			"resourceType": "Medication",
			"id": "med0312",
			"code": {
				"coding": [
					{
						"system": "http://snomed.info/sct",
						"code": "324689003",
						"display": "Nystatin 100,000 units/ml oral suspension (product)"
					}
				]
			}
		}
	],
	"identifier": [
		{
			"use": "official",
			"system": "http://www.bmc.nl/portal/prescriptions",
			"value": "12345689"
		}
	],
	"status": "completed",
	"intent": "order",
	"medicationReference": {
		"reference": "#med0312",
		"display": "Nystatin 100,000 u/ml oral suspension"
	}
}
```
3. Search with the following query: `MedicationRequest?medication.code=http://snomed.info/sct|324689003&_contained=true` 
4. See no results

**Expected behavior**
Should return the MedicationRequest

**Environment (please complete the following information):**
 - HAPI FHIR Version: 6.9.5
 - OS: MacOS
 - Browser: chrome

### PR DESCRIPTION
**Pull request by StevenXLi**
Closes #5307

### CODE REVIEW

**Review by lukedegruchy — APPROVED** [review]

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/7_0_0/5307-search-for-medicationrequests-with-medication-contained-does-not-return-correct-results.yaml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+type: fix
+issue: 5307
+title: "Previously, searching by medication in a chained search will not return the correct MedicationRequests if the
+medication is a contained resource. This is now fixed."
```
Comment (dmuylwyk) [inline hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/7_0_0/5307-search-for-medicationrequests-with-medication-contained-does-not-return-correct-results.yaml:5]: **suggestion (non-blocking):** This issue is broader in scope. I propose the following:

```suggestion
title: "Some search parameters include parenthetical expressions (e.g. `(MedicationRequest.medication as Reference)`). The leading `(` was causing searches using parenthetical expressions to fail where the reference target was a contained resource. This has been corrected."
```

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/QueryStack.java [hunk 2]
```diff
@@ -2808,7 +2808,7 @@ private boolean isReferenceParamValid(ReferenceParam theReferenceParam) {
 		private List<String> extractPaths(String theResourceType, RuntimeSearchParam theSearchParam) {
 			List<String> pathsForType = theSearchParam.getPathsSplit().stream()
 					.map(String::trim)
-					.filter(t -> t.startsWith(theResourceType))
+					.filter(t -> (t.startsWith("(") ? t.substring(1).trim() : t).startsWith(theResourceType))
```
Comment (dmuylwyk) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/QueryStack.java:2811]: **suggestion (blocking):** We don't need to modify the path in order to match it. This `trim` modifies the path such that the leading `(` is removed but the trailing `)` remains. I propose this alternative to preserve the encapsulated path:

```suggestion
					.filter(t -> t.startsWith(theResourceType) || t.startsWith("(" + theResourceType))
```

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4SearchContainedTest.java [hunk 3]
```diff
@@ -988,6 +998,47 @@ public void testContainedResourceParameterIsUsedInCache() {
 
 	}
 
+	@Test
+	public void testBundleReferenceSearchParam()  throws IOException {
```
Comment (dmuylwyk) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4SearchContainedTest.java:1002]: **suggestion (blocking):** This test name does not describe the test. I propose this alternative:

```suggestion
	public void testContainedSearchByTokenWithParentheticalExpression()  throws IOException {
```

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4SearchContainedTest.java [hunk 4]
```diff
@@ -988,6 +998,47 @@ public void testContainedResourceParameterIsUsedInCache() {
 
 	}
 
+	@Test
```
Comment (dmuylwyk) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4SearchContainedTest.java:1001]: **suggestion (blocking):** Please add a javadoc that identifies the ticket. This is super helpful to others in the future when they're trying to understand why we have multiple tests for contained search by token.

```suggestion
	/**
	 * See #5307
	 */
	@Test
```

**Review by dmuylwyk — APPROVED** [review]
Approved with some minor suggestions to clarify the nature of the bug fix, and to avoid potential future bugs as a result of modifying the path.

Nice work, @StevenXLi. Well caught!

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4SearchContainedTest.java [hunk 5]
```diff
@@ -988,6 +998,47 @@ public void testContainedResourceParameterIsUsedInCache() {
...
+
+			// -- update
+			mid1 = myMedicationRequestDao.create(medReq, mySrd).getId().toUnqualifiedVersionless();
+
+			MedicationRequest medReqCreated = myMedicationRequestDao.read(mid1);
+
+			ourLog.debug("Output: {}", myFhirContext.newJsonParser().setPrettyPrint(true).encodeResourceToString(medReqCreated));
+		}
+
+		String uri = myServerBase + "/MedicationRequest?medication.code=http://" + UrlUtil.escapeUrlParam("snomed.info/sct|324689003") + "&_contained=true";
```
[the remark this answers is not in the document]
Comment (StevenXLi) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4SearchContainedTest.java]: you're right, missed this, good catch

**Pull request merged**